File: bitmarket-common/src/main/java/com/bitwise/bitmarket/common/protocol/Offer.java
package com.bitwise.bitmarket.common.protocol;

import com.bitwise.bitmarket.common.PeerConnection;
import com.bitwise.bitmarket.common.currency.BtcAmount;
import com.bitwise.bitmarket.common.currency.FiatAmount;

public class Offer {

    private final OfferId id;
    private final int sequenceNumber;
    private final PeerId fromId;
    private final PeerConnection fromConnection;
    private final BtcAmount amount;
    private final FiatAmount bitcoinPrice;

    public Offer(
            OfferId id, int sequenceNumber, PeerId fromId, PeerConnection fromConnection,
            BtcAmount amount, FiatAmount bitcoinPrice) {
        this.id = id;
        this.sequenceNumber = sequenceNumber;
        this.fromId = fromId;
        this.fromConnection = fromConnection;
        this.amount = amount;
        this.bitcoinPrice = bitcoinPrice;
    }

    @Override
    public String toString() {
        return String.format("offer with ID %s from %s of %s (%s per BTC)",
                this.id.toString(),
                this.fromId.toString(),
                this.amount.toString(),
                this.bitcoinPrice.toString());
    }

    public OfferId getId() {
        return this.id;
    }

    public int getSequenceNumber() {
        return this.sequenceNumber;
    }

    public PeerId getFromId() {
        return this.fromId;
    }

    public PeerConnection getFromConnection() {
        return this.fromConnection;
    }

    public BtcAmount getAmount() {
        return this.amount;
    }

    public FiatAmount getBitcoinPrice() {
        return this.bitcoinPrice;
    }
}

File: bitmarket-common/src/main/scala/com/bitwise/bitmarket/common/protocol/Offer.scala
package com.bitwise.bitmarket.common.protocol

import com.bitwise.bitmarket.common.PeerConnection
import com.bitwise.bitmarket.common.currency.{FiatAmount, BtcAmount}
import com.bitwise.bitmarket.common.protorpc.{Callbacks, PeerSession}
import com.bitwise.bitmarket.common.protocol.protobuf.{
  BitmarketProtobuf => proto, ProtobufConversions}

case class Offer(
    id: String,
    sequenceNumber: Int,
    fromId: PeerId,
    fromConnection: PeerConnection,
    amount: BtcAmount,
    bitcoinPrice: FiatAmount) {

  override def toString = s"offer with ID $id from $fromId of $amount ($bitcoinPrice per BTC)"
}

object Offer {

  implicit val Write = new MessageSend[Offer] {

    def sendAsProto(msg: Offer, session: PeerSession) = {
      val stub = proto.PeerService.newStub(session.channel)
      stub.publish(
        session.controller,
        ProtobufConversions.toProtobuf(msg),
        Callbacks.noop[proto.PublishResponse])
    }
  }
}

